vpnglean/glean_parser_ext/cpp.py

```diff
@@ -0,0 +1,136 @@
+# -*- coding: utf-8 -*-
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+Outputter to generate C++ code for metrics.
+"""
+
+import json
+
+import jinja2
+from glean_parser import util
+from util import generate_metric_ids, generate_ping_ids, get_metrics
+
+
+def cpp_datatypes_filter(value):
+    """
+    A Jinja2 filter that renders C++ literals.
+
+    Based on Python's JSONEncoder, but overrides:
+      - lists to array literals {}
+      - strings to "value"
+    """
+
+    class CppEncoder(json.JSONEncoder):
+        def iterencode(self, value):
+            if isinstance(value, list):
+                yield "{"
+                first = True
+                for subvalue in list(value):
+                    if not first:
+                        yield ", "
+                    yield from self.iterencode(subvalue)
+                    first = False
+                yield "}"
+            elif isinstance(value, str):
+                yield '"' + value + '"'
+            else:
+                yield from super().iterencode(value)
+
+    return "".join(CppEncoder().iterencode(value))
+
+
+def type_name(obj):
+    """
+    Returns the C++ type to use for a given metric object.
+    """
+
+    if getattr(obj, "labeled", False):
+        class_name = util.Camelize(obj.type[8:])  # strips "labeled_" off the front.
+        return "Labeled<impl::{}Metric>".format(class_name)
+    generate_enums = getattr(obj, "_generate_enums", [])  # Extra Keys? Reasons?
+    if len(generate_enums):
+        for name, suffix in generate_enums:
+            if not len(getattr(obj, name)) and suffix == "Keys":
+                return util.Camelize(obj.type) + "Metric<NoExtraKeys>"
+            else:
+                # we always use the `extra` suffix,
+                # because we only expose the new event API
+                suffix = "Extra"
+                return "{}Metric<{}>".format(
+                    util.Camelize(obj.type), util.Camelize(obj.name) + suffix
+                )
+    return util.Camelize(obj.type) + "Metric"
+
+
+def extra_type_name(typ: str) -> str:
+    """
+    Returns the corresponding Rust type for event's extra key types.
+    """
+
+    if typ == "boolean":
+        return "bool"
+    elif typ == "string":
+        return "QString"
+    elif typ == "quantity":
+        return "int"
+    else:
+        return "UNSUPPORTED"
```

Comment:
Hum, I think this is here just to make Python return type checker happy (in case there is such a checker) or maybe for code readability? Because glean_parser does error if a random extra type is provided. I tried it and glean_parser gave me this:

```
/home/brizental/Projects/mozilla-vpn-client/glean/metrics.yaml:
    ```
    sample:
      addon_state_changed:
        extra_keys:
          state:
            type: lskajdhf
    ```

    'lskajdhf' is not one of ['string', 'boolean', 'quantity']
ERROR running glean_parser v6.2.0
❌ Found 1 errors.
```

Reply:
Oh, good point. `glean_parser` itself will catch that. This is only unsupported if glean_parser accepts (and your code uses) something that this glean_parser_ext doesn't know about.

Still probably good to explode codegen on it rather than relying on the compiler, though.